Libraries/LibWeb/Painting/PaintableBox.cpp

```diff
@@ -214,13 +214,86 @@ CSSPixelRect PaintableBox::compute_absolute_paint_rect() const
+    rect.set_y(absolute_rect.y() - box_model().padding.top);
+    rect.set_height(content_height() + box_model().padding.top + box_model().padding.bottom);
+    return rect;
+}
+
+CSSPixelRect PaintableBox::absolute_border_box_rect() const
+{
+    auto padded_rect = this->absolute_padding_box_rect();
+    CSSPixelRect rect;
+    auto use_collapsing_borders_model = override_borders_data().has_value();
+    // Implement the collapsing border model https://www.w3.org/TR/CSS22/tables.html#collapsing-borders.
+    auto border_top = use_collapsing_borders_model ? round(box_model().border.top / 2) : box_model().border.top;
+    auto border_bottom = use_collapsing_borders_model ? round(box_model().border.bottom / 2) : box_model().border.bottom;
+    auto border_left = use_collapsing_borders_model ? round(box_model().border.left / 2) : box_model().border.left;
+    auto border_right = use_collapsing_borders_model ? round(box_model().border.right / 2) : box_model().border.right;
+    rect.set_x(padded_rect.x() - border_left);
+    rect.set_width(padded_rect.width() + border_left + border_right);
+    rect.set_y(padded_rect.y() - border_top);
+    rect.set_height(padded_rect.height() + border_top + border_bottom);
+    return rect;
+}
+
 CSSPixelRect PaintableBox::absolute_paint_rect() const
 {
     if (!m_absolute_paint_rect.has_value())
         m_absolute_paint_rect = compute_absolute_paint_rect();
     return *m_absolute_paint_rect;
 }
 
+template<typename Callable>
+static CSSPixelRect united_rect_for_continuation_chain(PaintableBox const& start, Callable get_rect)
+{
+    // Combine the absolute rects of all paintable boxes of all nodes in the continuation chain. Without this, we
+    // calculate the wrong rect for inline nodes that were split because of block elements.
+    Optional<CSSPixelRect> result;
+
+    // FIXME: instead of walking the continuation chain in the layout tree, also keep track of this chain in the
+    //        painting tree so we can skip visiting the layout nodes altogether.
+    for (auto const* node = &start.layout_node_with_style_and_box_metrics(); node; node = node->continuation_of_node()) {
+        for (auto const& paintable : node->paintables()) {
+            if (!is<PaintableBox>(paintable))
+                continue;
+            auto const& paintable_box = static_cast<PaintableBox const&>(paintable);
+            auto paintable_border_box_rect = get_rect(paintable_box);
+            if (!result.has_value())
+                result = paintable_border_box_rect;
+            else if (!paintable_border_box_rect.is_empty())
+                result->unite(paintable_border_box_rect);
+        }
+    }
+    return result.value_or({});
+}
+
+CSSPixelRect PaintableBox::absolute_united_border_box_rect() const
+{
+    return united_rect_for_continuation_chain(*this, [](auto const& paintable_box) {
+        return paintable_box.absolute_border_box_rect();
+    });
+}
```

Comment:
hmm, shouldn't `absolute_border_box_rect()` behave like `absolute_united_border_box_rect()`. I would expect anything that does `dom_node->paintable_box()->absolute_border_box_rect()` to want united box by default.

Reply:
I've considered that, but that's not what you always want. Consider this fragment:

```html
foo<b>bar<div>baz</div></b>
```

"foo" should not be part of the paintable's rect, only "bar" and the `div` should be. At the moment I think the main purpose for the united versions of the rects is the inspector overlay; hit testing still requires the smaller rects.